File: CS/ExampleMainForm.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using DevExpress.XtraTreeList.Columns;
using DevExpress.XtraTreeList.Nodes;
using System.Threading;
using DevExpress.XtraTreeList;

namespace SimpleTreeListExample
{
    public partial class ExampleMainForm : Form
    {
        public ExampleMainForm()
        {
            InitializeComponent();
        }

        private void Form1_Load(object sender, EventArgs e)
        {

            TreeListColumn newColumnDescription = MyTtreeList.Columns.Add();
            newColumnDescription.FieldName = "Description";
            newColumnDescription.VisibleIndex = 0;

            MyTtreeList.SelectImageList = imageListForTree;
            MyTtreeList.DataSource = new object();
        }

        private void treeList1_VirtualTreeGetChildNodes(object sender, DevExpress.XtraTreeList.VirtualTreeGetChildNodesInfo e)
        {
            String sCurrentNode = e.Node as String;
            if (sCurrentNode == null)
            {
                e.Children = new string[] { "Root node 1", 
                                            "Root node 2", 
                                            "Root node 3" };
            }
            else
            {
                e.Children = CreateChildrenNodes(sCurrentNode);
            }
        }

        private BindingList<string> CreateChildrenNodes(string sCurrentNode)
        {
            BindingList<string> sChildren = new BindingList<string>();
            using (BackgroundWorker workerForChildren = new BackgroundWorker())
            {
                workerForChildren.DoWork += new DoWorkEventHandler(workerForChildren_DoWork);
                workerForChildren.RunWorkerAsync(sChildren);
            }
            return sChildren;
        }

        void workerForChildren_DoWork(object sender, DoWorkEventArgs e)
        {
            BindingList<string> sChildren = e.Argument as BindingList<string>;
            for (int i = 0; i < 4; i++)
            {
                BeginInvoke(new MethodInvoker(delegate() {                
                    sChildren.Add(String.Format("Child node {0}", i)); }));
                Thread.Sleep(1000);
            }
        }

        private void treeList1_VirtualTreeGetCellValue(object sender, DevExpress.XtraTreeList.VirtualTreeGetCellValueInfo e)
        {
            e.CellData = e.Node;
        }

        private void treeList1_GetSelectImage(object sender, DevExpress.XtraTreeList.GetSelectImageEventArgs e)
        {
            if (e.Node.Id < imageListForTree.Images.Count)
            {
                e.NodeImageIndex = e.Node.Id;
            }
            else
            {
                e.NodeImageIndex = -1;
            }
        }
    }
}

File: VB/ExampleMainForm.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.XtraTreeList.Columns
Imports System.Threading
Imports DevExpress.XtraTreeList

Namespace SimpleTreeListExample

    Public Partial Class ExampleMainForm
        Inherits Form

        Public Sub New()
            InitializeComponent()
        End Sub

        Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs)
            Dim newColumnDescription As TreeListColumn = MyTtreeList.Columns.Add()
            newColumnDescription.FieldName = "Description"
            newColumnDescription.VisibleIndex = 0
            MyTtreeList.SelectImageList = imageListForTree
            MyTtreeList.DataSource = New Object()
        End Sub

        Private Sub treeList1_VirtualTreeGetChildNodes(ByVal sender As Object, ByVal e As VirtualTreeGetChildNodesInfo)
            Dim sCurrentNode As String = TryCast(e.Node, String)
            If Equals(sCurrentNode, Nothing) Then
                e.Children = New String() {"Root node 1", "Root node 2", "Root node 3"}
            Else
                e.Children = CreateChildrenNodes(sCurrentNode)
            End If
        End Sub

        Private Function CreateChildrenNodes(ByVal sCurrentNode As String) As BindingList(Of String)
            Dim sChildren As BindingList(Of String) = New BindingList(Of String)()
            Using workerForChildren As BackgroundWorker = New BackgroundWorker()
                AddHandler workerForChildren.DoWork, New DoWorkEventHandler(AddressOf workerForChildren_DoWork)
                workerForChildren.RunWorkerAsync(sChildren)
            End Using

            Return sChildren
        End Function

        Private Sub workerForChildren_DoWork(ByVal sender As Object, ByVal e As DoWorkEventArgs)
            Dim sChildren As BindingList(Of String) = TryCast(e.Argument, BindingList(Of String))
            For i As Integer = 0 To 4 - 1
                BeginInvoke(New MethodInvoker(Sub() sChildren.Add(String.Format("Child node {0}", i))))
                Thread.Sleep(1000)
            Next
        End Sub

        Private Sub treeList1_VirtualTreeGetCellValue(ByVal sender As Object, ByVal e As VirtualTreeGetCellValueInfo)
            e.CellData = e.Node
        End Sub

        Private Sub treeList1_GetSelectImage(ByVal sender As Object, ByVal e As GetSelectImageEventArgs)
            If e.Node.Id < imageListForTree.Images.Count Then
                e.NodeImageIndex = e.Node.Id
            Else
                e.NodeImageIndex = -1
            End If
        End Sub
    End Class
End Namespace
